### PR TITLE
issue#1436 [Enhancement]: Unknown system variable 'lc_time_names'

### PR DESCRIPTION
**Pull request by bobbird1995**
### Task Description
[Enhancement]: Unknown system variable 'lc_time_names'. https://github.com/oceanbase/oceanbase/issues/1436
https://github.com/oceanbase/oceanbase/issues/1436

### Solution Description
OB currently does not have a locale related variable configuration like MySQL, which is' lc_time_names'. The value set should be of String type. lc_time_names will affect the output results of date related expressions, including:
-DATE_ FORMAT()
-DAYNAME()
-MONTHNAME()
-FROM_UNIXTIME()

In addition to adding variables to OB, I also made modifications to the affected expressions mentioned above.

but, I only implemented these locale ， such as en_US, zh_CN, zh_TW, ja_JP, ko_KR.

### Passed Regressions

I have put the test cases to the “.tools\deploy\mysql_test\test_suite\type_date\t\”

### Upgrade Compatibility

I added a new variable，so this commit should be merged into version 4.3.

### CODE REVIEW

File: deps/oblib/src/lib/locale/ob_locale.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+*          http://license.coscl.org.cn/MulanPubL-2.0
+* THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
+* EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
+* MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
+* See the Mulan PubL v2 for more details.
+*/
+#include "lib/locale/ob_locale_type.h"
+namespace oceanbase {
+namespace common {
+// 注意ob_locae_ar_AE的number刚好是6，要放在ob_locales中的第6位，用于定位
```
Comment (akaError) [inline deps/oblib/src/lib/locale/ob_locale.cpp:15]: comment in English

File: deps/oblib/src/lib/locale/ob_locale.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    if(NULL == ob_locales[i]){
+      continue;
+    } else {
+      if(0 == locale_name.case_compare(ob_locales[i]->name)){
+        result_locale = ob_locales[i];
+        break;
+      }
+    }
+  }
+  if( offset == i){
```
Comment (akaError) [inline deps/oblib/src/lib/locale/ob_locale.cpp:70]: why not use (NULL == result_locale) , offset == I is curious

File: deps/oblib/src/lib/locale/ob_locale.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  for( ; i < offset ; i++){
+    if(NULL == ob_locales[i]){
+      continue;
+    } else {
+      if(0 == locale_name.case_compare(ob_locales[i]->name)){
+        break;
+      }
+    }
+  }
+  if( offset == i){
```
Comment (akaError) [inline deps/oblib/src/lib/locale/ob_locale.cpp:89]: ){ should add one more white space -> ") {", please check all the format

File: deps/oblib/src/lib/locale/ob_locale.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  }
+  return result_locale;
+}
+
+bool is_valid_ob_locale(const ObString &locale_name) {
+  bool ret = true;
+  int64_t offset = sizeof(ob_locales)/sizeof(OB_LOCALE *)-1;
+  int64_t i = 0;;
+  for( ; i < offset ; i++){
+    if(NULL == ob_locales[i]){
```
Comment (akaError) [inline deps/oblib/src/lib/locale/ob_locale.cpp:81]: format

File: deps/oblib/src/lib/locale/ob_locale_type.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+ public:
+  unsigned int number;
+  const char *name;
+  const char *description;
+  const bool is_ascii;
+  OB_LOCALE_TYPE *month_names;
+  OB_LOCALE_TYPE *ab_month_names;
+  OB_LOCALE_TYPE *day_names;
+  OB_LOCALE_TYPE *ab_day_names;
+  unsigned int  max_month_name_length;
```
Comment (akaError) [inline deps/oblib/src/lib/locale/ob_locale_type.h:50]: member variables end with '_', such as month_names -> "month_names_"

File: src/sql/engine/expr/ob_expr_time.cpp [hunk 6]
```diff
@@ -218,10 +220,18 @@ int ObExprTimeBase::calc(const ObExpr &expr, ObEvalCtx &ctx, ObDatum &expr_datum
       } else {
         int idx = ot.parts_[type] - 1;
         if(0 <= idx  && idx < 7) {
-          const ObString &name = daynames[idx];
-          if (OB_FAIL(ObExprUtil::set_expr_ascii_result(expr, ctx, expr_datum, name))) {
-            LOG_WARN("failed to exec set_expr_ascii_result", K(expr), K(ctx), K(expr_datum), K(name));
-          }          
+           // const ObString &name = daynames[idx];
+          if(OB_FAIL(session->get_locale_name(locale_name))){
```
Comment (akaError) [inline src/sql/engine/expr/ob_expr_time.cpp:224]: format

File: src/sql/engine/expr/ob_expr_time.cpp [hunk 7]
```diff
@@ -233,9 +243,17 @@ int ObExprTimeBase::calc(const ObExpr &expr, ObEvalCtx &ctx, ObDatum &expr_datum
       } else {
         int idx = ot.parts_[type] - 1;
         if(0 <= idx  && idx < 12) {
-          const ObString &month_name = monthnames[idx];
-          if (OB_FAIL(ObExprUtil::set_expr_ascii_result(expr, ctx, expr_datum, month_name))) {
-            LOG_WARN("failed to exec set_expr_ascii_result", K(expr), K(ctx), K(expr_datum), K(month_name));
+          // const ObString &month_name = monthnames[idx];
```
Comment (akaError) [inline src/sql/engine/expr/ob_expr_time.cpp:246]: remove unused line

File: src/sql/engine/expr/ob_expr_time.cpp [hunk 8]
```diff
@@ -218,10 +220,18 @@ int ObExprTimeBase::calc(const ObExpr &expr, ObEvalCtx &ctx, ObDatum &expr_datum
...
-          if (OB_FAIL(ObExprUtil::set_expr_ascii_result(expr, ctx, expr_datum, name))) {
-            LOG_WARN("failed to exec set_expr_ascii_result", K(expr), K(ctx), K(expr_datum), K(name));
-          }          
+           // const ObString &name = daynames[idx];
+          if(OB_FAIL(session->get_locale_name(locale_name))){
+              LOG_WARN("failed to get locale time name", K(expr), K(expr_datum));
+          } else {
+            OB_LOCALE *ob_cur_locale = ob_locale_by_name(locale_name);
+            OB_LOCALE_TYPE *locale_type = ob_cur_locale->day_names;
+            const char ** locale_daynames = locale_type->type_names_;
```
Comment (akaError) [inline src/sql/engine/expr/ob_expr_time.cpp:229]: in here do not split mysql mode with oracle, does u sure it is right in oceanbase oracle mode?

Comment (bobbird1995) [reply src/sql/engine/expr/ob_expr_time.cpp:229]: > in here do not split mysql mode with oracle, does u sure it is right in oceanbase oracle mode?

Are you talking about method “ob_locale_by_name”？

File: deps/oblib/src/lib/locale/ob_locale.cpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                            &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US,
+                            &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US,
+                            &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US,
+                            &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US,
+                            &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US, NULL};
+
+OB_LOCALE *ob_locale_by_number(unsigned int number) {
+  OB_LOCALE *locale = NULL;
+  if (number >= MONTH_LENGTH - 1) {
+    //do nothing
```
Comment (akaError) [inline deps/oblib/src/lib/locale/ob_locale.cpp:48]: it seems 'ob_locale_by_number' not use now ? and if number >= MONTH_LENGTH - 1 , if finally return NULL? it is a  reasonable return value ? at least should return a en_us?

File: deps/oblib/src/lib/locale/ob_locale.cpp [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+* EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
+* MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
+* See the Mulan PubL v2 for more details.
+*/
+#include "lib/locale/ob_locale_type.h"
+namespace oceanbase {
+namespace common {
+// Note that the index of the ob_locale_ja_JP is exactly 2 and should be placed in the 2nd position in the ob_locales.
+OB_LOCALE *ob_locales[] = {&ob_locale_en_US, &ob_locale_en_US, &ob_locale_ja_JP, &ob_locale_en_US,
+                            &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US, &ob_locale_en_US,
```
Comment (akaError) [inline deps/oblib/src/lib/locale/ob_locale.cpp:17]: why set all is en_US? this is wrong way

File: deps/oblib/src/lib/locale/ob_locale_type.cc [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+                          &ob_locale_type_ab_day_names_en_US, 9, 9,
+                          '.',        /* decimal point en_US */
+                          ',',        /* thousands_sep en_US */
+                          "\x03\x03"/* grouping      en_US */);
+
+
+
+/* 1.需要修改成ob前缀，把最后一个换成null
+* 2.长度需要用定义好的宏MONTH_LENGTH和DAY_LENGTH
+*3. OB_LOCALE中没有err_msg的结构
```
Comment (akaError) [inline deps/oblib/src/lib/locale/ob_locale_type.cc:47]: remove useless Chinese comment

File: src/share/system_variable/ob_system_variable.cpp [hunk 12]
```diff
@@ -2389,6 +2390,28 @@ int ObSysVarOnCheckFuncs::check_runtime_filter_type_is_valid(
   return ret;
 }
 
+
```
Comment (akaError) [inline src/share/system_variable/ob_system_variable.cpp:2393]: one one empty line is enough

File: src/share/system_variable/ob_system_variable.cpp [hunk 13]
```diff
@@ -2389,6 +2390,28 @@ int ObSysVarOnCheckFuncs::check_runtime_filter_type_is_valid(
...
+  if (true == set_var.is_set_default_) {
+    //do nothing
+  } else if (!is_valid_ob_locale(locale_val)) {            //check if the variable is valid
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("invalid locale type", K(ret), K(in_val));
+  }
+  return ret;
+}
+
+
```
Comment (akaError) [inline src/share/system_variable/ob_system_variable.cpp:2413]: only one empty line is enough

File: src/sql/session/ob_basic_session_info.cpp [hunk 14]
```diff
@@ -3459,6 +3459,21 @@ int ObBasicSessionInfo::get_init_connect(ObString &str) const
...
+{
+  int ret = OB_SUCCESS;
+  if (lib::is_mysql_mode()) {
+    if(OB_FAIL(get_string_sys_var(SYS_VAR_LC_TIME_NAMES, str))) {
+      ret = OB_ERR_UNEXPECTED;
+      LOG_WARN("failed to load sys variables", "var_name",SYS_VAR_LC_TIME_NAMES, K(ret));
+    }
+   } else {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_ERROR("oracle mode does not support lc_time_names", K(lib::is_oracle_mode()), K(ret));
```
Comment (akaError) [inline src/sql/session/ob_basic_session_info.cpp:3472]: LOG_WARN, not use LOG_ERROR

File: src/share/system_variable/ob_system_variable_factory.cpp [hunk 15]
```diff
@@ -5907,6 +5920,17 @@ int ObSysVarFactory::create_sys_var(ObIAllocator &allocator_, ObSysVarClassType
...
       break;
     }
+    case SYS_VAR_LC_TIME_NAMES: {
+      void *ptr = NULL;
+      if (OB_ISNULL(ptr = allocator_.alloc(sizeof(ObSysVarLcTimeNames)))) {
+        ret = OB_ALLOCATE_MEMORY_FAILED;
+        LOG_ERROR("fail to alloc memory", K(ret), K(sizeof(ObSysVarLcTimeNames)));
+      } else if (OB_ISNULL(sys_var_ptr = new (ptr)ObSysVarLcTimeNames())) {
+        ret = OB_ALLOCATE_MEMORY_FAILED;
+        LOG_ERROR("fail to new ObSysVarLcTimeNames", K(ret));
```
Comment (akaError) [inline src/share/system_variable/ob_system_variable_factory.cpp:5930]: log_warn, not log_error

**Pull request closed without merging**